File: works/16928.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Pair
{
    int cur, cnt;
    public Pair(int _cur, int _cnt) { cur = _cur; cnt = _cnt; }
}

public class Main
{
    static int n, m;
    static boolean[] mark = new boolean[101];
    static int[] moveTo = new int[101];
    static int answer;

    static void bfs()
    {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(1, 0));

        while(!q.isEmpty())
        {
            Pair top = q.poll();
            int cur = top.cur;
            int cnt = top.cnt;

            if(cur == 100) { answer = cnt; break; }

            for(int i = 1; i <= 6; i++)
            {
                int ncur = cur + i;
                if(ncur > 100 || mark[ncur]) continue;
                if(moveTo[ncur] != 0)
                {
                    ncur = moveTo[ncur];
                    mark[ncur] = true;
                    q.add(new Pair(ncur, cnt+1));
                }
                else
                {
                    mark[ncur] = true;
                    q.add(new Pair(ncur, cnt+1));
                }
            }
        }
    }

    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] _s = br.readLine().split(" ");
        n = Integer.parseInt(_s[0]);
        m = Integer.parseInt(_s[1]);
        for(int i = 0; i < n+m; i++)
        {
            _s = br.readLine().split(" ");
            moveTo[Integer.parseInt(_s[0])] = Integer.parseInt(_s[1]);
        }

        bfs();
        System.out.print(answer);
    }

}

File: works/16928.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class Pair(_cur:Int, _cnt:Int)
{
    var cur = _cur
    var cnt = _cnt
}

var n = 0
var m = 0
var mark = Array(101) { false }
var moveTo = Array(101) { 0 }
var answer = 0

fun bfs()
{
    var q = LinkedList<Pair>()
    q.add(Pair(1,0))

    while(!q.isEmpty())
    {
        var top = q.poll()
        var cur = top.cur
        var cnt = top.cnt

        if(cur == 100) { answer = cnt; break }

        for(i in 1 .. 6)
        {
            var ncur = cur + i
            if(ncur > 100 || mark[ncur]) continue
            if(moveTo[ncur] != 0)
            {
                ncur = moveTo[ncur]
                mark[ncur] = true
                q.add( Pair(ncur, cnt+1))
            }
            else
            {
                mark[ncur] = true
                q.add( Pair(ncur, cnt+1))
            }
        }
    }
}

fun main()
{
    val br = BufferedReader(InputStreamReader(System.`in`))
    var _s = br.readLine().split(' ')
    n = _s[0].toInt()
    m = _s[1].toInt()
    for(i in 0 until n+m)
    {
        _s = br.readLine().split(' ')
        moveTo[_s[0].toInt()] = _s[1].toInt()
    }

    bfs()
    print(answer)
}